src/widgets/sessions.ts

```diff
@@ -0,0 +1,131 @@
+// SessionManager exposes `JupyterLab.serviceManager.sessions` to user python kernel
+
+import { SessionManager } from '@jupyterlab/services';
+import { ISerializers, WidgetModel } from '@jupyter-widgets/base';
+import { toArray } from '@lumino/algorithm';
+import { FocusTracker } from '@lumino/widgets';
+import { MODULE_NAME, MODULE_VERSION } from '../version';
+import { Session } from '@jupyterlab/services';
+
+/**
+ * The model for a Session Manager
+ */
+export class SessionManagerModel extends WidgetModel {
+  /**
+   * The default attributes.
+   */
+  defaults(): any {
+    return {
+      ...super.defaults(),
+      _model_name: SessionManagerModel.model_name,
+      _model_module: SessionManagerModel.model_module,
+      _model_module_version: SessionManagerModel.model_module_version,
+      current_session: null,
+      sessions: [],
+    };
+  }
+
+  /**
+   * Initialize a CommandRegistryModel instance.
+   *
+   * @param attributes The base attributes.
+   * @param options The initialization options.
+   */
+  initialize(attributes: any, options: any): void {
+    this._sessions = SessionManagerModel.sessions;
+    this._tracker = SessionManagerModel.tracker;
+    super.initialize(attributes, options);
+    this.on('msg:custom', this._onMessage.bind(this));
+    this._sessions.runningChanged.connect(this._sendSessions, this);
+    this._tracker.currentChanged.connect(this._currentChanged, this);
+    this._tracker.activeChanged.connect(this._currentChanged, this);
+    this._sendSessions();
+    this._sendCurrent();
+  }
+
+  /**
+   * Handle a custom message from the backend.
+   *
+   * @param msg The message to handle.
+   */
+  private _onMessage(msg: any): void {
+    switch (msg.func) {
+      case 'get_current':
+        this._sendCurrent();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * get sessionContext from a given widget instance
+   *
+   * @param widget widget tracked by app.shell._track (FocusTracker)
+   */
+  private _getSessionContext(widget: any): Session.IModel | {} {
+    if (widget?.sessionContext) {
+      // covers both ConsolePanel and NotebookPanle
+      if (widget.sessionContext.session?.model) {
+        return widget.sessionContext.session.model;
+      }
+    }
+    return {}; // empty object serializes to empty dict in python
```

Comment:
It looks like this could be condensed into:

```suggestion
    return widget?.sessionContext?.session?.model ?? {};
```

Reply:
Absolutely. Still trying to get a hang of the typescript magic. let me change that too.